[RFC] hub dashboard orchestration (#5)
reichlab/decisions: decisions/2024-12-13-rfc-hub-dashboard-orchestration.md

1. a modeling hub that contains model output data
2. a dashboard repository that defines parameters for the site and the dashboard

The admins of at least one hubverse project has expressed a desire to not manage
any more workflows.

### Aims

 - define a method for generating a hubverse dashboard website using the three
   main components
 - define places where the outputs for each of the three components should live
 - provide a solution that does not significantly complicate the workflows of
   hub admins
 - provide a solution that does not add significant maintenance burden for the
   hubverse core team

### Anti-Aims

 - TBD

## Decision

 - We will use [re-useable GitHub
   workflows](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#limitations)
   to orchestrate each of the build process---these workflows can be reused in
   any repository.
 - We will generate artifacts from each workflow and store them as separate
   orphan branches in the dashboard repository.
 - We will build an experimental GitHub App that allows hub admins to
   optionally register their repository to build the dashboard using our
   centralized workflows.

>  - We will host the App's webhook code on a free service like glitch.io

Reviewer: Repeating my question from a Zoom call: our synchronous discussion was helpful for understanding how the webhook/API component of this decision improves usability for hub admins while maintaining a good security posture.

That said, a lot of us have pushed back against running and maintaining an API/web service, as it adds a non-trivial moving part to our setup.

As expected, everyone has their own .02 about where to land on the hub admin usability <--> hub dev maintainability spectrum. I'm wondering how much (if at all) the choice of host/language for the API impacts that perspective.

Author: > I'm wondering how much (if at all) the choice of host/language for the API impacts that perspective.

I suspect that a language for the API that is more aligned with the lab languages would make the app option more appealing (depending on how robust the abstraction is). Most of the challenge with the app code is dealing with GitHub's API: https://github.com/hubverse-org/hub-dashboard-control-room/blob/60fc0660a96f054123d702c7dbe4b2108545f73f/app/index.js#L4-L24

### Other Options Considered

> #### A set of GitHub workflows copied to individual hub dashboard repos
> 
> Similar to current Hubverse-developed GitHub actions. Hub admins copy workflow
> and update as needed. Configured to run on a cron schedule and manually as
> needed.

Reviewer 2: I understood the original language here as saying that the full workflow files would be copied over to the individual hub dashboard repos.  This suggestion revises the text to talk about reusable workflows.

```suggestion
#### A set of GitHub workflows copied to individual hub dashboard repos calling centrally-maintained reusable workflows

Similar to current Hubverse-developed GitHub actions, but putting most of
the heavy lifting in centrally-maintained reusable workflows. Hub admins copy
a (relatively small) workflow to their dashboard repo and update it as needed.
Configured to run on a cron schedule and manually as needed.
```

Author: I believe this is the correct interpretation. It is not to say that we did not consider copying over full workflows (instead of reusable workflows), but this is a more accurate representation.

- PRO: Doesn't require a security token that has write access to a control
room-type repository
- PRO: Doesn't require API/web service
- CON: Increased maintenance burden for hub admins
- CON: Potential increased support burden for hub devs
- NEUTRAL: Does not use GitHub app

> #### Centralized workflows invoked by individual hub repos via PAT

Reviewer 2: This option comes closest to what I understood as the option we chose yesterday to focus on in a next phase of exploration.  But I'm confused by the part of this title that says "invoked by individual hub repos via PAT".  I was thinking there was a way this could be done without requiring individual hub admins to maintain a PAT for calling the re-usable workflows.  This would involve placing a reduced version of [the build.yaml file[(==](https://github.com/hubverse-org/hub-dashboard-control-room/blob/main/.github/workflows/build.yaml) in each hub dashboard repo, calling the centrally-maintained re-usable workflow(s).  I thought that would not require a PAT, but am I missing something?

Author: No. This was the one I was vehemently against because of the PAT. The one above this is what we asked me to create an example of (I walked you through this yesterday https://github.com/zkamvar/hub-dashboard-workflow/blob/23af1fcc2c1631f3be4afcba0746401971a40a53/.github/workflows/rebuild-data.yaml)

This particular workflow was one that was proposed where the workflow would send a `repository_dispatch` event to the central GitHub repository. The problem with this is that it requires a PAT that can only be generated by us, which provides a security risk.

Reviewer 2: Thanks for the clarification; that makes sense.  I was confused because I understood the one above as indicating that the "full" workflows would need to be copied to each repo, which is bad.  I'll resolve this conversation and make a suggestion above attempting to clarify.

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. Hub admins generate and maintain a

> personal GitHub Personal Access token with write access to the control room repo.

Author: This is incorrect. It's not possible for hub admins to generate a PAT to the control room repo. Only administrators for the hubverse and/or owners of that control room repository can generate this PAT.

Author: ```suggestion
room workflows that perform dashboard tasks. Hub admins are provided a
personal GitHub Personal Access token with write access to the control room repo. Hubverse organization owners are required to maintain and rotate the token.
```

- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: Decentralized secrets management
- NEUTRAL: Does not use GitHub app

#### Centralized workflows invoked by individual hub repos via GitHub App token

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. Hub admins store the GitHub app
token as repo secret and use it when invoking the control room workflows.

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: Sharing the GitHub App token
- NEUTRAL: Uses a GitHub app

#### Centralized workflows automatically copied to individual dashboard repos

Dashboard workflows are maintained in the control room app, but some kind of
CI/build process automatically adds (or updates) them to individual hub
dashboard repos.

- PRO: Benefits of centralized repo + sidesteps need for hub dashboard repos
to have write access to control room app
- PRO: Doesn't require API/web service
- CON: In practice, keeping things synchronized between individual repos and

> a central repo is difficult

Author: To expand a bit on this, it is difficult not because the process involved is challenging per-se, but it requires folks to be careful about the permissions of the tokens they use. In order to update workflow files, a GitHub token needs both the `contents: write` and `workflows: write` contexts.

I've done this before and it's stressful: https://github.com/carpentries/sandpaper/tree/main/inst/workflows#updates

Reviewer 2: added a suggested comment below to get the reason for this difficulty into the document

- NEUTRAL: Does not use GitHub app

#### Secrets vault

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. The boilerplate workflow retrieves
the GitHub app secret from a secrets value (e.g., Hashicorp Vault or AWS
Secrets Manager).

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)

> - CON: It replaces the API/web service moving part with a different moving part

Reviewer: A way this could work ... GitHub actions in the hub dashboard repos could authenticate to a vault app via its GitHub OIDC provider and get the GitHub app token ([HasiCorp Vault](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault) or [AWS](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) for example). OIDC is how the [AWS Upload action](https://github.com/hubverse-org/hubverse-actions/blob/main/hubverse-aws-upload/hubverse-aws-upload.yaml#L38) is able to push data to a hub's S3 bucket.

As with getting hubs onto the Hubverse-hosted AWS cloud, there would need to be an onboarding process for hub dashboards. Essentially, we'd have to make sure each individual hubverse dashboard repo is allowed to access the vault, either manually or via infrastructure as code (the [relevant bit for AWS cloud onboarding](https://github.com/hubverse-org/hubverse-infrastructure/blob/main/src/hubverse_infrastructure/hubs/iam.py#L11))

This _might_ require less ongoing maintenance than running an API, but it's still not trivial.

Author: Four things:

1. there would be no need for an app here because the secret is a PAT for the control-room repository that's needed to send a repository dispatch event, not an app secret. 
2. we lose control of the PAT as soon as it leaves the vault.
3. every solution that involves sharing a PAT or some sort of secret is effectively re-inventing one aspect of what an app is doing in a much less secure way. 
4. there's the ongoing maintenance of rotation of the token for security.

Reviewer 3: It seems like some of the conversation here is hinging on security concerns. It's not fully clear to me what the concerns are, or what the implications would be if the token was either misused or fell into "bad actor" hands. What could get borked? What is a "worst case scenario" here?

Author: > What could get borked? What is a "worst case scenario" here?

This token would be specific for the control-room repository. It would allow anyone with access to that token to insert _any code they want_ into the this central repository. They could insert code that would insert malicious javascript in the dashboard websites so that it either redirects to another page (either adware, scam, or other malicious site), adds a popup asking for credentials, or any other number of attacks that can be pulled off with javascript.

Reviewer 2: added a suggestion for a new point in the "Aims" section clarifying the importance of security

# 2024-12-13 Orchestration of the Dashboard Build Process

## Context

[The hubverse dashboard](../project-posters/hub-dashboard/hub-dashboard.md) is
a project that generates a website with auto-generated dashboards for forecasts
and step-ahead forecast visualizations.

This project has three main components:

1. static site generator
2. forecast visualization generator
3. evaluations dashboard generator

This project requires two data sources:

1. a modeling hub that contains model output data
2. a dashboard repository that defines parameters for the site and the dashboard

> The admins of at least one hubverse project has expressed a desire to not manage

Reviewer 3: has --> have

> The admins of at least one hubverse project has expressed a desire to not manage

Reviewer 3: ```suggestion
The admins of at least one hubverse project have expressed a desire to not manage
```

any more workflows.

### Aims

>  - define a method for generating a hubverse dashboard website using the three
>    main components

Reviewer 3: ```suggestion
 - define a method for generating a hubverse dashboard website using all or a subset of the three
   main components
```

 - define places where the outputs for each of the three components should live
 - provide a solution that does not significantly complicate the workflows of
   hub admins
 - provide a solution that does not add significant maintenance burden for the
   hubverse core team

> 

Reviewer 2: ```suggestion
- Provide a solution that is secure. For instance, a solution without proper attention to security could allow bad actors to insert code that would insert malicious javascript in the dashboard websites.

```

### Anti-Aims

 - TBD

## Decision

 - We will use [re-useable GitHub
   workflows](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#limitations)
   to orchestrate each of the build process---these workflows can be reused in
   any repository.
 - We will generate artifacts from each workflow and store them as separate
   orphan branches in the dashboard repository.
 - We will build an experimental GitHub App that allows hub admins to
   optionally register their repository to build the dashboard using our
   centralized workflows.
 - We will host the App's webhook code on a free service like glitch.io

> 

Reviewer 2: Adding this to reflect where we landed on this.  We currently have both the app described in the immediately preceding points as well as the idea in the new point here.
```suggestion
- We will also build out a method in which each hub dashboard repo has relatively-lightweight workflows that call centrally-maintained reusable workflows

```

Reviewer 3: I might suggest saying something explicit like a final decision about whether or not to pick one of the App v. centrally maintained workflows has not be made yet.

Author: @nickreich, I agree with this and I will add the context.

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: Decentralized secrets management
- NEUTRAL: Does not use GitHub app

#### Centralized workflows invoked by individual hub repos via GitHub App token

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. Hub admins store the GitHub app
token as repo secret and use it when invoking the control room workflows.

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: Sharing the GitHub App token
- NEUTRAL: Uses a GitHub app

#### Centralized workflows automatically copied to individual dashboard repos

Dashboard workflows are maintained in the control room app, but some kind of
CI/build process automatically adds (or updates) them to individual hub
dashboard repos.

- PRO: Benefits of centralized repo + sidesteps need for hub dashboard repos
to have write access to control room app
- PRO: Doesn't require API/web service

> - CON: In practice, keeping things synchronized between individual repos and
> a central repo is difficult

Reviewer 2: ```suggestion
- CON: In practice, keeping things synchronized between individual repos and
a central repo is difficult because it requires careful attention to the permissions of GitHub tokens.
```

> # 2024-12-13 Orchestration of the Dashboard Build Process

Reviewer 2: Although it's valuable to think this through in the specific context of the dashboard build process, I've been wondering if it would (or would not) be helpful to expand the scope of discussion here a bit to the general problem of workflow management in the hubverse.  For example, Zhian and Anna were scrambling a bit last week to get workflow updates for validations out to hubs, when the `ubuntu-latest` runner changed the major OS version.  We also have several other actions used for other purposes.  It seems like it would be nice to arrive at a general solution.

Reviewer: For posterity, repeating my zoom comments here.

I agree that the multitude of workflows is a challenge to tackle, but we should focus on solving that problem in the context of the hub dashboard project. If we arrive at a solution that's generalizable to the hubverse as a whole, we can start a new stream of work to implement it on a wider scale.

Author: Agreed with @bsweger, the general workflow management problem is out of scope here.

The admins of at least one hubverse project has expressed a desire to not manage
any more workflows.

### Aims

 - define a method for generating a hubverse dashboard website using the three
   main components
 - define places where the outputs for each of the three components should live
 - provide a solution that does not significantly complicate the workflows of
   hub admins
 - provide a solution that does not add significant maintenance burden for the
   hubverse core team

### Anti-Aims

 - TBD

## Decision

 - We will use [re-useable GitHub
   workflows](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#limitations)
   to orchestrate each of the build process---these workflows can be reused in
   any repository.
 - We will generate artifacts from each workflow and store them as separate
   orphan branches in the dashboard repository.
 - We will build an experimental GitHub App that allows hub admins to
   optionally register their repository to build the dashboard using our
   centralized workflows.
 - We will host the App's webhook code on a free service like glitch.io

> ### Other Options Considered

Reviewer 2: I've been wondering what other options might be available.  Here are a few thoughts, I'm not sure how viable any of them are.

- Would it be possible to use a github app to manage permissions to enable pushing workflow updates to hub dashboard repos (and/or hub repos)?
- Could we use reusable workflows more heavily, so that in each hub dashboard repo (or hub repo) there is only a small/minimal workflow that will need minimal updating, and which calls a single reusable workflow that is maintained centrally in 1 place in the hubverse? The idea is that this would make there be less "surface area" for failure in workflows in hub or hub dashboard repos, so that they are less likely to need updates.
- Could we use Docker in workflows to get a more stable run environment?  Again, with the goal of reducing the frequency with which we need to make updates.

There may be other ideas as well!

>  - A set of GitHub workflows only

Reviewer 2: I've been wondering what other options might be available.  Here are a few thoughts, I'm not sure how viable any of them are.

- Would it be possible to use a github app to manage permissions to enable pushing workflow updates to hub dashboard repos (and/or hub repos)?
- Could we use reusable workflows more heavily, so that in each hub dashboard repo (or hub repo) there is only a small/minimal workflow that will need minimal updating, and which calls a single reusable workflow that is maintained centrally in 1 place in the hubverse? The idea is that this would make there be less "surface area" for failure in workflows in hub or hub dashboard repos, so that they are less likely to need updates.
- Could we use Docker in workflows to get a more stable run environment?  Again, with the goal of reducing the frequency with which we need to make updates.

There may be other ideas as well!

>  - A set of GitHub workflows only

Reviewer: ```suggestion
#### A set of GitHub workflows copied to individual hub dashboard repos

Similar to current Hubverse-developed GitHub actions. Hub admins copy workflow
and update as needed. Configured to run on a cron schedule and manually as
needed.

- PRO: Doesn't require a security token that has write access to a control
room-type repository
- PRO: Doesn't require API/web service
- CON: Increased maintenance burden for hub admins
- CON: Potential increased support burden for hub devs
- NEUTRAL: Does not use GitHub app

#### Centralized workflows invoked by individual hub repos via PAT

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. Hub admins generate and maintain a
personal GitHub Personal Access token with write access to the control room repo.

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: Decentralized secrets management
- NEUTRAL: Does not use GitHub app

#### Centralized workflows invoked by individual hub repos via GitHub App token

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. Hub admins store the GitHub app
token as repo secret and use it when invoking the control room workflows.

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: Sharing the GitHub App token
- NEUTRAL: Uses a GitHub app

#### Centralized workflows automatically copied to individual dashboard repos

Dashboard workflows are maintained in the control room app, but some kind of
CI/build process automatically adds (or updates) them to individual hub
dashboard repos.

- PRO: Benefits of centralized repo + sidesteps need for hub dashboard repos
to have write access to control room app
- PRO: Doesn't require API/web service
- CON: In practice, keeping things synchronized between individual repos and
a central repo is difficult
- NEUTRAL: Does not use GitHub app

#### Secrets vault

Hub dashboard repos have a small boilerplate workflow that invokes the control
room workflows that perform dashboard tasks. The boilerplate workflow retrieves
the GitHub app secret from a secrets value (e.g., Hashicorp Vault or AWS
Secrets Manager).

- PRO: Centralized workflows are easier for hub devs to maintain
- PRO: Doesn't require API/web service
- PRO: Less workflow boilerplate for hub admins (but not zero)
- CON: It replaces the API/web service moving part with a different moving part
(albeit one that's similar to the existing onboarding process for cloud-based
hubs)
- NEUTRAL: Uses a GitHub app
```

>    - **Not chosen because of maintenance burden for hubverse admins** (see 

Reviewer: > We will host the App's webhook code on a free service like glitch.io

Assuming that decision not to rely solely on GitHub workflows is related to the quoted text from the above list of decisions...

I'm still unclear on why we need to host webhook code on Glitch, etc. instead of having a manually-triggered action in the hub-specific repos (e.g., [variant-nowcast-hub-dashboard](https://github.com/reichlab/variant-nowcast-hub-dashboard)) that use the GitHub api to trigger a [workflow](https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event)/[repository](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-a-repository-dispatch-event) dispatch event in the [central dashboard "control room"](https://github.com/hubverse-org/hub-dashboard-control-room/tree/main/.github/workflows)

Is there a technical and/or code-reuse reason we haven't tried on-demand updates w/o writing our own intervening API?

If not, it's appealing to remove the extra moving part (i.e., Glitch/server). Asking hub admins to manually run a GitHub workflow to update the dashboard doesn't seem overly-burdensome, especially for a first iteration of dashboards.

Reviewer: Adding that I also saw this note above:

> The admins of at least one hubverse project has expressed a desire to not manage
any more workflows.

If there's a way to include a generalizable workflow in the dashboard template (i.e., admins don't have to make changes to it) or if there's a way to share the hub-specific version of the repo, that doesn't seem terrible.

In any case, I'd like to make sure we're making a good tradeoff between user desire for "no more workflows" and the added complexity introduced by introducing another service into the dashboard update process.

Author: > If there's a way to include a generalizable workflow in the dashboard template (i.e., admins don't have to make changes to it).

As soon as a workflow is introduced, it's out of our control. We can do our darnedest to make sure admins don't have to make changes to it, but I have yet to ever find an example where that's the case.

> if there's a way to share the hub-specific version of the repo, that doesn't seem terrible.

I'm not sure I understand this. Can you elaborate?
